From 8cd164170a494a9eef7e179f7782f404bf059c44 Mon Sep 17 00:00:00 2001
From: Du Chenyang <chenyang.du@intel.com>
Date: Thu, 14 Mar 2013 11:42:20 +0800
Subject: [PORT FROM MAIN] [PATCH] OpenSSL: Fix race condition issue with tlsext_ecpointformatlist

BZ: 58183

In multi-thread situations, s->session->tlsext_ecpointformatlist is not
thread safe without lock's protection.

Change-Id: I89bf6a24c4a33c7845e45f3bf1dbf08d32b730da
Orig-Change-Id: Ideef43b11ddf649b662b46a5bcb40e8ed300372f
Signed-off-by: Du Chenyang <chenyang.du@intel.com>
Reviewed-on: http://android.intel.com:8080/96497
Reviewed-by: Shuster, Norvin <norvin.shuster@intel.com>
Reviewed-by: Allyn, Mark A <mark.a.allyn@intel.com>
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Shen, JianchunX <jianchunx.shen@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 ssl/t1_lib.c |   21 ++++++++++++++-------
 1 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/ssl/t1_lib.c b/ssl/t1_lib.c
index 28d45b3..3d2b391 100644
--- a/ssl/t1_lib.c
+++ b/ssl/t1_lib.c
@@ -1469,15 +1469,22 @@ int ssl_parse_serverhello_tlsext(SSL *s, unsigned char **p, unsigned char *d, in
 				*al = TLS1_AD_DECODE_ERROR;
 				return 0;
 				}
-			s->session->tlsext_ecpointformatlist_length = 0;
-			if (s->session->tlsext_ecpointformatlist != NULL) OPENSSL_free(s->session->tlsext_ecpointformatlist);
-			if ((s->session->tlsext_ecpointformatlist = OPENSSL_malloc(ecpointformatlist_length)) == NULL)
+			if (!s->hit)
 				{
-				*al = TLS1_AD_INTERNAL_ERROR;
-				return 0;
+				if(s->session->tlsext_ecpointformatlist)
+					{
+					OPENSSL_free(s->session->tlsext_ecpointformatlist);
+					s->session->tlsext_ecpointformatlist = NULL;
+					}
+				s->session->tlsext_ecpointformatlist_length = 0;
+				if ((s->session->tlsext_ecpointformatlist = OPENSSL_malloc(ecpointformatlist_length)) == NULL)
+					{
+					*al = TLS1_AD_INTERNAL_ERROR;
+					return 0;
+					}
+				s->session->tlsext_ecpointformatlist_length = ecpointformatlist_length;
+				memcpy(s->session->tlsext_ecpointformatlist, sdata, ecpointformatlist_length);
 				}
-			s->session->tlsext_ecpointformatlist_length = ecpointformatlist_length;
-			memcpy(s->session->tlsext_ecpointformatlist, sdata, ecpointformatlist_length);
 #if 0
 			fprintf(stderr,"ssl_parse_serverhello_tlsext s->session->tlsext_ecpointformatlist ");
 			sdata = s->session->tlsext_ecpointformatlist;
-- 
1.7.4.1

